contrib/DC-GAN.ipynb

```diff
@@ -0,0 +1,569 @@
+    "        # Reset gradients\n",
+    "        optimizer.zero_grad()\n",
+    "\n",
+    "        # 1. Train on Real Data\n",
+    "        prediction_real = discriminator(real_data)\n",
+    "        # Calculate error and backpropagate\n",
+    "        error_real = loss(prediction_real, real_data_target(real_data.size(0), device))\n",
+    "        \n",
+    "\n",
+    "        # 2. Train on Fake Data\n",
+    "        prediction_fake = discriminator(fake_data)\n",
+    "        # Calculate error and backpropagate\n",
+    "\n",
+    "        error_fake = loss(prediction_fake, fake_data_target(real_data.size(0), device))\n",
+    "        total_error = error_real + error_fake\n",
+    "        total_error.backward()\n",
+    "\n",
+    "        # Update weights with gradients\n",
+    "        xm.optimizer_step(optimizer)\n",
+    "\n",
+    "        return total_error, prediction_real, prediction_fake\n",
+    "        #return (0, 0, 0)\n",
+    "\n",
+    "    def train_step_generator(optimizer, fake_data, device):\n",
+    "        # Reset gradients\n",
+    "        optimizer.zero_grad()\n",
+    "        # Sample noise and generate fake data\n",
+    "        prediction = discriminator(fake_data)\n",
+    "        # Calculate error and backpropagate\n",
+    "        error = loss(prediction, real_data_target(prediction.size(0), device))\n",
+    "        error.backward()\n",
+    "        # Update weights with gradients\n",
+    "        xm.optimizer_step(optimizer)\n",
+    "\n",
+    "        # Return error\n",
+    "        return error\n",
+    "\n",
+    "    # Notice the use of .detach() when fake_data is to passed into discriminator\n",
+    "    def train_loop_fn(loader):\n",
+    "        tracker = xm.RateTracker()\n",
+    "        for n_batch, (real_batch,_) in enumerate(loader):\n",
+    "            # Train Step Descriminator\n",
+    "            real_data = Variable(real_batch).to(device)\n",
+    "            fake_data = generator(noise(real_data.size(0), device))\n",
+    "            d_error, d_pred_real, d_pred_fake = train_step_discriminator(d_optimizer,\n",
+    "                                                                real_data, fake_data.detach(), device)\n",
+    "            #Train Step Generator\n",
+    "            fake_data = generator(noise(real_batch.size(0), device))\n",
+    "            g_error = train_step_generator(g_optimizer, fake_data, device)\n",
+    "            #print(f'D_ERROR: {d_error.item()}, G_ERROR: {g_error.item()}')\n",
+    "        return d_error.item(), g_error.item()\n",
+    "\n",
+    "\n",
+    "    for epoch in range(1, FLAGS['num_epochs'] +1):\n",
+    "        para_loader = pl.ParallelLoader(train_loader, [device])\n",
+    "        d_error, g_error = train_loop_fn (para_loader.per_device_loader(device))\n",
+    "        xm.master_print(\"Finished training epoch {}: D_error:{}, G_error: {}\".format(epoch, d_error, g_error))\n",
+    "        if epoch %5 == 0:\n",
```

Comment:
nit: `epoch % 5` (missing space)

Reply:
Moved plot call outside the loop. Since I haven't figured out a way to dynamically show the images from within a thread. Other option would be to store each of them with a uniq name.